### PR TITLE
[LANG-1662] Let ReflectionToStringBuilder only reflect given field names

### PR DESCRIPTION
**Pull request by GutoVeronezi**
As commented in JIRA issue [LANG-1662](https://issues.apache.org/jira/browse/LANG-1662)  and the mailing list ([here](https://lists.apache.org/thread/348sxhc277r45hkwt39rpdwxmxt5lrrs) and [here](https://lists.apache.org/thread/4nxrlfrzssv5w09c2gs9sphn7rgoc613)), the class **ReflectionToStringBuilder** has methods to exclude fields from *toString*; If we have an object with several fields and want to reflect only some, we have to list all the fields that we don't want to reflect and pass to **excludeFieldNames**, or use the `ToStringBuilder` and call the method `append` with the keys and values.

To exemplify what I mean:

- In a case where we need to print only the fields `a`,  `b`, and `e` from the class `UseCase`, the current solution is:

```java
       Example example = new UseCase.Example();
       example.setA("test");
       example.setB(1);
       example.setC(1.1);
       example.setD(true);
       example.setE(false);
       example.setF(false);
       example.setG(true);

       ToStringBuilder toStringBuilder = new ToStringBuilder(example,
       ToStringStyle.JSON_STYLE);
       toStringBuilder.append("a", example.getA());
       toStringBuilder.append("b", example.getB());
       toStringBuilder.append("e", example.isE());
       System.out.println(toStringBuilder.build());
```

With these changes, the solution would be:

```java
       Example example = new UseCase.Example();
       example.setA("test");
       example.setB(1);
       example.setC(1.1);
       example.setD(true);
       example.setE(false);
       example.setF(false);
       example.setG(true);
     
       ReflectionToStringBuilder reflectionToStringBuilder = new ReflectionToStringBuilder(example, ToStringStyle.JSON_STYLE);
       reflectionToStringBuilder.setSelectedFieldNames("a", "b", "e");
       System.out.println(reflectionToStringBuilder.build());
```

Both ways would generate the result `{"a":"test","b":1,"e":false}`, but there are some negative points while using `ToStringBuilder`:   
- We instantiate the `ToStringBuilder` with the object, but it's meaningless to the situation, as we need to set every key/value;
- We can put whatever we want as attribute name and value, therefore it's not related to the object we want to print;
- It's verbose; The more attributes we want to print, the more calls to append we need;

With the proposed solution, one has only to pass the field names as a parameter to reflect it.

### CODE REVIEW

File: src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java
Comment (rafaelweingartner) [inline src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:394]: in case of Null parameter, What is the output:

Comment (rafaelweingartner) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:394]: I see. Can you add these details in the JavaDocs? When reading I got the impression that Null would return an empty string.

**Review by rafaelweingartner — APPROVED** [review]

**Comment by GutoVeronezi** [comment]
Could anyone review this PR?

**Comment by garydgregory** [comment]
I will review later in the week.

**Comment by GutoVeronezi** [comment]
Thanks, @garydgregory.

**Comment by garydgregory** [comment]
Hello @GutoVeronezi 

Thank you for your PR.

The name `setSelectedFieldNames1 is confusing because it is not symmetric (for lack of a better word) with `setExcludeFieldNames`, I think it would be best as `setIncludeFieldNames` which then is a bit odd since ALL fields are included by default, so there should be _clear_ documentation that using `setIncludeFieldNames` overrides this default behavior.

The following behavior should throw an `IllegalStateException` IMO:
```
       ReflectionToStringBuilder reflectionToStringBuilder = new ReflectionToStringBuilder(example, ToStringStyle.JSON_STYLE);
       reflectionToStringBuilder.setIncludedFieldNames("a", "b", "e");
       reflectionToStringBuilder.setExcludedFieldNames("a", "b", "e"); // throws IllegalStateException?
       System.out.println(reflectionToStringBuilder.build()); // throws IllegalStateException?
```
IOW what does it mean to set a field as both included and excluded, does the "last API call win" or, does the code simply throw an `IllegalStateException`  if the include and exclude set intersect?

**Comment by GutoVeronezi** [comment]
Hi, @garydgregory, 

Thanks for the review, I'll address the changes regarding the feature name (`setSelectedFieldNames` -> `setIncludeFieldNames`) as well the documentation of its use.

---

Regarding the intersection between `included` and `excluded`, I see it as a complement rather than an intersection;  The current behavior of this class, as you said, is to include all fields by default. If we have a class with fields `a`, `b` and `c` and instantiate and reflect it, both of them will be reflected. If we set them in the excluded field names, the reflection will return empty:

```java
import org.apache.commons.lang3.builder.ReflectionToStringBuilder;
import org.apache.commons.lang3.builder.ToStringStyle;

public class Test {

    private String a;
    private String b;
    private String c;

    public Test(String a, String b, String c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }

    public static void main(String[] args) {
        Test test = new Test("this is field 'a'", "this is field 'b'", "this is field 'c'");
        ReflectionToStringBuilder reflection = new ReflectionToStringBuilder(test, ToStringStyle.JSON_STYLE);
        System.out.println(String.format("Before excluding: %s", reflection.build()));

        reflection = new ReflectionToStringBuilder(test, ToStringStyle.JSON_STYLE); // Recreating the object because once builded it caches the result.
        reflection.setExcludeFieldNames("a", "b", "c");
        System.out.println(String.format("After excluding: %s", reflection.build()));
    }
```

Result:

```
Before excluding: {"a":"this is field 'a'","b":"this is field 'b'","c":"this is field 'c'"}
After excluding: {}
```

In this example we're creating a complement of `excluded` (fields `a`, `b` and `c`) regarding set `included` (the default fields `a`, `b` and `c`), which will result in an empty set, or, in other words, nothing to reflect. The behavior per se will be the same, however, with this proposal, we'll be able to override the `included` set. Therefore, I think instead of throwing an exception, we should add an explanation in the documentation about it and let the users decide how to handle it.

If this is proposal is not valid and we proceed with the exception proposal, which cases should we consider to throw it? `included` and `excluded` are equal or just by having one field of `included` in `excluded` should throw an exception?

**Comment by garydgregory** [comment]
Hi @GutoVeronezi 
You did not answer my question. See my code example; it should be one of your unit tests. I do not know (yet) what would be reasonable as the behavior. 

@kinow , any thoughts?

**Review by kinow — COMMENTED** [review]
I think the feature can be added, as it would make sense to have both exclude & include methods. I also agree with @garydgregory , allowing including and excluding the same fields sound like something that would create confusion or unwanted bugs in the code of users.

I think ideally we would throw an exception for illegal argument for both include/exclude when there are conflicts.

Also, the javadocs use "on output", I think it sounds a little off. I'm not a native speaker, but I think "in the output" sounds better. Nit-picking :grimacing: rest of the code looks OK to me, we just need to handle the case highlighted by Gary, and update the tests :+1: 

Thanks @GutoVeronezi !
Bruno

File: src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java
Comment (kinow) [inline src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:494]: s/will be include/will be included

**Review by kinow — CHANGES_REQUESTED** [review]
Hi,

I think the code looks OK, but there's something wrong when mixing values and `null`'s. I think it's normal as the PR is trying to let include & exclude fields to co-exist, hopefully not too hard to fix it.

First, I think the unit test written is really good! Testing that fields and values exist, and also that the omitted fields and values do not exist in the `toString` :clap: 

@GutoVeronezi , I think we need more tests, including the absurd tests. I had some spare time to review this one, so I decided to hack the unit test and adding a few more to play with the test. Here's one that **worked**:

```java
// Not really a test, I was first looking at the output...
    @Test
    public void test () {
        ReflectionToStringBuilder builder = new ReflectionToStringBuilder(new TestFeature());
        builder.setExcludeFieldNames("field1");
        builder.setIncludeFieldNames(null);
        System.out.println(builder.toString());
    }

// prints: org.apache.commons.lang3.builder.ReflectionToStringBuilderIncludeTest$TestFeature@76a3e297[field2=value 2,field3=value 3,field4=value 4,field5=value 5]
```

Now this one fails (I simply added one more `null` value):

```java
    @Test
    public void test () {
        ReflectionToStringBuilder builder = new ReflectionToStringBuilder(new TestFeature());
        builder.setExcludeFieldNames("field1");
        builder.setIncludeFieldNames(null, null);
        System.out.println(builder.toString());
    }
```

The code above results in:

```
java.lang.IllegalStateException: Fields [] set in "includeFieldsNames" were set in "excludeFieldNames" too.
	at org.apache.commons.lang3.builder.ReflectionToStringBuilder.accept(ReflectionToStringBuilder.java:658)
	at org.apache.commons.lang3.builder.ReflectionToStringBuilder.appendFieldsIn(ReflectionToStringBuilder.java:699)
	at org.apache.commons.lang3.builder.ReflectionToStringBuilder.toString(ReflectionToStringBuilder.java:915)
	at org.apache.commons.lang3.builder.ReflectionToStringBuilderIncludeTest.test(ReflectionToStringBuilderIncludeTest.java:188)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)

```

Take a look and see if we can add more of these random tests, with empty arrays, a combination of invalid fields in the include and exclude, etc. We might end up with a lot more tests, but I think that's necessary if we are to allow for include & exclude filters simultaneously.

Cheers
Bruno

**Comment by GutoVeronezi** [comment]
Hi, @kinow,

Thanks for the review.

I fixed the bug with the exception and added more unit tests. I couldn't think in many other situations to validate; therefore, if you think in other use cases, just let me know and I create the tests.

Best regards,
Daniel

**Review by kinow — APPROVED** [review]
Looks good to me, addresses the case described in JIRA, and I can't think of more tests :slightly_smiling_face: 

Probably good somebody else/others to have a look too.

Thanks @GutoVeronezi !!
Bruno

**Comment by GutoVeronezi** [comment]
Hi, can anyone else review this PR?

Best regards,
Daniel

**Comment by garydgregory** [comment]
Applied PR and:
- Added missing @since tags to new public and protected methods and
fields
- Fixed Javadocs
- Reuse own ArrayUtils API for validation
- Move methods to AB order

**Comment by GutoVeronezi** [comment]
@garydgregory, I have a question, why close the PR and push the changes as if they were yours instead of merging the PR? I know that you referenced my PR in the commit message and so on, but for anybody that looks at the git history, the credits are yours. If you think that changes/amendments were necessary, you could have asked, and I would haply apply them. However, this way seems a bit odd. At least, quite different from the other OpenSource interactions I have had so far. Is this a common practice here in the commons-lang community?

**Comment by kinow** [comment]
Hi @GutoVeronezi 

I had a quick look via GitHub UI and it is not showing you as co-author. When a maintainer needs to perform changes during the merge, it is common practice here in Commons (also in other OSS projects), to squash/rebase/edit commits.

I think maybe @garydgregory was trying that but accidentally replaced authorship? Normally both original author & Commons maintainer appear as the authors of the commits. We can edit that commit's authorship & push-force if necessary, but let's wait for @garydgregory 's reply to confirm if that was really the case :+1: 

Thanks!
Bruno

**Comment by GutoVeronezi** [comment]
Hi, @kinow, thanks for the reply.

@garydgregory, it feels a bit weird for us here on the other side proposing things to contribute back to commons-lang and interacting with reviewers to see somebody "taking over the changes". If the PR is missing something or there is a better way to do something, I would expect the reviewers to point what can be improved, or at least it being merged with some changes/rebase and so on (as described by @kinow https://github.com/apache/commons-lang/pull/849#issuecomment-1175649542). This way of applying the changes (closing the PR and pushing it as yours, without co-authoring) discourages new contributors, because we do not receive the credit for all the work we did (neither in Git history nor in GitHub Insights).

Anyways, thanks for accepting my changes, I will now work on our side to use the new release of commons-lang as soon as it is released.

**Pull request closed without merging**
